### PR TITLE
Display 'Search globally' as the last sharees list element

### LINKED ISSUE #3847: Query lookup server
Query lookup search needs to be opt-in as it is now, but we should enhance the way:

Desktop it is not clearly visible how to do it
image

Discussed with @jancborchardt @nimisha-vijay 
image

- show "search globally" as last search "result"
- after click on it
  -  change "search globally" to "global search results"
  - append search results from lookup server
    - if none are found, show "no results found"

### PR DESCRIPTION
**Pull request by allexzander**
this will fix #3847

### CODE REVIEW

File: src/gui/filedetails/shareemodel.cpp [hunk 1]
```diff
@@ -233,9 +261,48 @@ void ShareeModel::shareesFetched(const QJsonDocument &reply)
...
+    std::sort(std::begin(_sharees), std::end(_sharees), [](const auto &left, const auto &right) {
+        const auto isRightElementLookupPlaceholder = right->type() == Sharee::LookupServerSearch
+            || right->type() == Sharee::LookupServerSearchResults;
+
+        const auto isLeftElementLookupPlaceholder = left->type() == Sharee::LookupServerSearch
+            || left->type() == Sharee::LookupServerSearchResults;
+
+        return !isLeftElementLookupPlaceholder && isRightElementLookupPlaceholder;
+    });
 }
```
Comment (claucambra) [inline src/gui/filedetails/shareemodel.cpp:306]: If we need to sort the elements in the model this should be done in a separate `QSortFilterProxyModel` (with testing for the correct order), and through an override of `QSortFilterProxyModel::lessThan`

File: src/gui/filedetails/shareemodel.cpp [hunk 2]
```diff
@@ -275,4 +342,13 @@ void ShareeModel::filterSharees()
     Q_EMIT shareesReady();
 }
 
+void ShareeModel::slotDarkModeChanged()
+{
+    for (int i = 0; i < _sharees.size(); ++i) {
+        if (_sharees[i]->updateIconUrl()) {
+            Q_EMIT dataChanged(index(i), index(i), {IconRole});
+        }
+    }
```
Comment (claucambra) [inline src/gui/filedetails/shareemodel.cpp:351]: Why not just call `dataChanged` once for all the rows in the model?

File: src/gui/filedetails/sharemodel.cpp [hunk 3]
```diff
@@ -939,6 +939,8 @@ void ShareModel::createNewUserGroupShareWithPassword(const ShareePtr &sharee, co
         return;
     }
 
+    emit startedCretingNewUserGroupShare();
```
Comment (claucambra) [inline src/gui/filedetails/sharemodel.cpp:942]: nitpick: spelling

File: src/gui/filedetails/sharemodel.h [hunk 4]
```diff
@@ -116,6 +116,7 @@ class ShareModel : public QAbstractListModel
     void passwordSetError(const QString &shareId, const int code, const QString &message);
     void requestPasswordForLinkShare();
     void requestPasswordForEmailSharee(const OCC::ShareePtr &sharee);
+    void startedCretingNewUserGroupShare() const;
```
Comment (claucambra) [inline src/gui/filedetails/sharemodel.h:119]: nitpick: spelling

File: src/gui/sharee.h [hunk 5]
```diff
@@ -35,35 +35,47 @@ Q_DECLARE_LOGGING_CATEGORY(lcSharing)
...
 
-    explicit Sharee(const QString shareWith,
-        const QString displayName,
-        const Type type);
+    Q_REQUIRED_RESULT QString format() const;
+    Q_REQUIRED_RESULT QString shareWith() const;
+    Q_REQUIRED_RESULT QString displayName() const;
+    Q_REQUIRED_RESULT QString iconUrl() const;
+    Q_REQUIRED_RESULT QString iconUrlColoured() const;
+    Q_REQUIRED_RESULT Type type() const;
```
Comment (claucambra) [inline src/gui/sharee.h:57]: We should avoid using `Q_REQUIRED_RESULT` as this is not documented in Qt and seems to be a private thing -- better use `[[nodiscard]]`

File: src/gui/filedetails/shareemodel.cpp (hunk 2 above)
Comment (allexzander) [reply src/gui/filedetails/shareemodel.cpp:351]: @claucambra I was not able to come up with a proper call as per your suggestion. Do you have an example? Whatever I try fails to update the view as a result.

File: src/gui/filedetails/ShareeSearchField.qml [hunk 6]
```diff
@@ -219,13 +223,25 @@ TextField {
                     anchors.left: parent.left
                     anchors.right: parent.right
 
+                    enabled: model.type !== Sharee.LookupServerSearchResults
+                    hoverEnabled: model.type !== Sharee.LookupServerSearchResults
```
[the remark this answers is not in the document]
Comment (allexzander) [reply src/gui/filedetails/ShareeSearchField.qml:227]: @claucambra yes, when the display text turns to "Global search results" or "No search results" as per issue's plan, we should not allow clicking on this item as it does nothing and is merely there for information (maybe not the best design but what Tobias wants)

**Review by claucambra — APPROVED** [review]

**Pull request merged**